Sources/SwiftSyntax/SourceLocation.swift

```diff
@@ -0,0 +1,344 @@
+//===--------------- SourceLocation.swift - Source Locations --------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+/// Represents a source location in a Swift file.
+public struct SourceLocation: Codable {
+  /// The line in the file where this location resides. 1-based.
+  public let line: Int
+
+  /// The UTF-8 byte offset from the beginning of the line where this location
```

Comment:
If we're gonna grapheme-break while computing these, we should change this comment. Otherwise, we should be strict about making sure this is actually a UTF-8 offset.

Reply:
If this is a UTF-8 offset, then we’ll need to make sure to add `String(char).utf8.count`, whereas if it’s supposed to be user-visible columns (where 👨‍👨‍👧‍👦 takes up one visible column), this shouldn’t be UTF-8.

Reply:
It's still unclear to me, where to add `String(char).utf8.count`, at the absolute-offset -> line/col conversion or vice-versa, and where is `char` coming from ?

Are you suggesting this because the `column` is "1-based" ? Would it be clearer as a UTF8-based column if it was actually "0-based" ?

Reply:
Sorry, wasn’t clear. Yes I do mean in the AbsolutePosition -> line/column conversion, and the `char` is coming from the other discussion in the comments about how best to accumulate source lengths.

Reply:
The code as-is does take into account the length of the new-line character. `position(ofLine:, column:)` derives a correct UTF8 offset from the beginning of the line, no matter what was the utf8 count of the new-line, AFAIK (then adds `1` to make it "1-based").

Reply:
I’m starting to think this should not be a UTF-8 offset at all. Column, as a 1-indexed value, is much more for human comprehension than indexing, so using UTF-8 for this pessimizes lines with multi-byte characters.

Maybe this shouldn’t be UTF-8, and should only be the number of `Character`s since the previous newline?

Reply:
Determining the column characters is way more expensive though. Once we determine the line, getting the UTF8 offset from beginning of line is O(1), while getting the character index would involve doing another walk of the syntax tree, or creating a data structure to efficiently represent all utf8 character indices.

It's not clear to me the complexity and expense is worth it, because clients:
* Most likely have the the source string and can do the any conversion they want themselves. Determining the line number can assist in making the conversion more efficient.
* They probably have to deal with swift/sourcekitd providing UTF8 column offsets as well.

I'm fine with making line/column 0-based as a more natural fit for UTF8 columns.

(BTW, to make sure it's clear, FWIW I did not change the interpretation of `SourceLocation.column` it was always UTF8-based.)

Reply:
Okay, that’s fair, then. Let’s keep it as-is, I must have been mistaken about the previous behavior. Thanks!